shared/network_linux.go

```diff
@@ -1,15 +1,999 @@
+
+	netnsID := C.netns_get_nsid(C.__s32(f.Fd()))
+	if netnsID < 0 {
+		return nil, fmt.Errorf("Failed to retrieve network namespace id")
+	}
+
+	ret := C.netns_getifaddrs(&ifaddrs, netnsID)
+	if ret < 0 {
+		return nil, fmt.Errorf("Failed to retrieve network interfaces and addresses")
+	}
+	defer C.netns_freeifaddrs(ifaddrs)
+
+	// We're using the interface name as key here but we should really
+	// switch to the ifindex at some point to handle ip aliasing correctly.
+	networks := map[string]api.NetworkState{}
+
+	for addr := ifaddrs; addr != nil; addr = addr.ifa_next {
+		var address [C.INET6_ADDRSTRLEN]C.char
+		addNetwork, networkExists := networks[C.GoString(addr.ifa_name)]
+		if !networkExists {
+			addNetwork = api.NetworkState{
+				Addresses: []api.NetworkStateAddress{},
+				Counters:  api.NetworkStateCounters{},
+			}
+		}
+
+		if addr.ifa_addr.sa_family == C.AF_INET || addr.ifa_addr.sa_family == C.AF_INET6 {
+			netState := "down"
+			netType := "unknown"
+
+			if (addr.ifa_flags & C.IFF_BROADCAST) > 0 {
+				netType = "broadcast"
+			}
+
+			if (addr.ifa_flags & C.IFF_LOOPBACK) > 0 {
+				netType = "loopback"
+			}
+
+			if (addr.ifa_flags & C.IFF_POINTOPOINT) > 0 {
+				netType = "point-to-point"
+			}
+
+			if (addr.ifa_flags & C.IFF_UP) > 0 {
+				netState = "up"
+			}
+
+			family := "inet"
+			if addr.ifa_addr.sa_family == C.AF_INET6 {
+				family = "inet6"
+			}
+
+			addr_ptr := C.get_addr_ptr(addr.ifa_addr)
+			if addr_ptr == nil {
+				return nil, fmt.Errorf("Failed to retrieve valid address pointer")
+			}
+
+			address_str := C.inet_ntop(C.int(addr.ifa_addr.sa_family), addr_ptr, &address[0], C.INET6_ADDRSTRLEN)
+			if address_str == nil {
+				return nil, fmt.Errorf("Failed to retrieve address string")
+			}
+
+			if addNetwork.Addresses == nil {
+				addNetwork.Addresses = []api.NetworkStateAddress{}
+			}
+
+			goAddrString := C.GoString(address_str)
+			scope := "global"
+			if strings.HasPrefix(goAddrString, "127") {
+				scope = "local"
+			}
+
+			if goAddrString == "::1" {
+				scope = "local"
+			}
+
+			if strings.HasPrefix(goAddrString, "169.254") {
+				scope = "link"
+			}
+
+			if strings.HasPrefix(goAddrString, "fe80:") {
+				scope = "link"
+			}
+
+			address := api.NetworkStateAddress{}
+			address.Family = family
+			address.Address = goAddrString
+			address.Netmask = fmt.Sprintf("%d", int(addr.ifa_prefixlen))
+			address.Scope = scope
+
+			addNetwork.Addresses = append(addNetwork.Addresses, address)
+			addNetwork.State = netState
+			addNetwork.Type = netType
+			addNetwork.Mtu = int(addr.ifa_mtu)
+		} else if addr.ifa_addr.sa_family == C.AF_PACKET {
+			var buf [1024]C.char
+
+			hwaddr := C.get_packet_address(addr.ifa_addr, &buf[0], 1024)
+			if hwaddr == nil {
+				return nil, fmt.Errorf("Failed to retrieve hardware address")
+			}
+
+			addNetwork.Hwaddr = C.GoString(hwaddr)
+		}
+
+		ifName := C.GoString(addr.ifa_name)
+		if !networkExists {
+			addNetwork.Counters = NetworkGetCounters(ifName)
```

Comment:
This can't work, /proc/net/dev only shows devices in your current network namespace and we run this from another network namespace.

Reply:
Ugh, namespaced, right. IFLA_STATS it is then.

Reply:
Fixed, this uses IFLA_STATS now.